Add chain_color option to graph.curve.params() for customizing MCMC line color

Kwan-Jenny opened the pull request:
This PR introduces a new `chain_color` argument to the `graph.curve.params()` function, allowing users to control whether MCMC sample lines are colored by chain (default) or plotted in black (`chain_color = FALSE`).

### Summary of changes:
- **New feature**: `chain_color` argument added to `graph.curve.params()`.
-  Default behavior retains color-coding by chain; setting `chain_color = FALSE` makes all MCMC lines black.
-  **Examples updated** in `inst/examples/exm-graph.curve.params.R` to demonstrate:
  - Default behavior
  - Custom quantiles
  - Use of `chain_color = FALSE`
- **Tests updated** in `tests/testthat/test-graph.curve.params.R`:
  - Visual regression tests using `vdiffr::expect_doppelganger()`
  - Includes plots with default settings, custom quantiles, and black chain coloring
  - Also tests `log_x = TRUE` option for coverage

This update enhances visual flexibility and clarity for plotting antibody decay models, especially in publication-quality figures with many MCMC samples.

Kwan-Jenny committed:
# Refactor graph.curve.params(): numeric quantiles, remove show_quantiles

Kwan-Jenny committed:
Resolved merge conflicts from main

Merge branch 'main' into update_graph_curve2

# Conflicts:
#	R/graph.curve.params.R
#	man/graph.curve.params.Rd
#	tests/testthat/_snaps/graph.curve.params/curve-samples.svg
#	tests/testthat/test-graph.curve.params.R

Kwan-Jenny committed:
Add chain_color option to toggle MCMC line coloring by chain or black.

Kwan-Jenny committed:
# update example file and test file

Kwan-Jenny committed:
devtools::document

Kwan-Jenny committed:
update version

Kwan-Jenny committed:
# revise lint problem

Kwan-Jenny committed:
# update NEWS.md file and devtools:document()

Kwan-Jenny committed:
# add test svg files

Kwan-Jenny committed:
#update version

Kwan-Jenny committed:
Merge branch 'main' into update_graph_curve2

Kwan-Jenny committed:
# test(snapshots): Update visual snapshots

Kwan-Jenny committed:
Merge remote-tracking branch 'origin/update_graph_curve2'

Kwan-Jenny committed:
# feat: Improve legend titles in curve graph

# Conditionally sets the legend title in `graph.curve.params()` to accurately reflect the plot contents.

- Displays "MCMC chain" when colored chains and quantiles are shown.
- Hides the title when black chains and quantiles are shown to avoid confusion.
- Displays "Quantile" when only quantile lines are plotted.

d-morrison commented on R/graph.curve.params.R:
```diff
@@ -14,6 +14,9 @@
 #' (default, `TRUE`) or linear scale (`FALSE`)?
+#' @param chain_color logical, if `TRUE` (default), MCMC chain lines
+#' are colored by chain.
+#' If `FALSE`, all MCMC chain lines are black.
```
```suggestion
#' @param chain_color [logical]: if [TRUE] (default), MCMC chain lines
#' are colored by chain.
#' If [FALSE], all MCMC chain lines are black.
```

d-morrison commented on tests/testthat/test-graph.curve.params.R:
```diff
@@ -49,7 +57,7 @@ test_that(
       curve,
       n_curves = Inf,
-      show_quantiles = TRUE,
+      quantiles = NULL,
```
why remove quantiles from this test?

Kwan-Jenny replied:
@d-morrison This specific test was designed to verify that the function works correctly when a user intentionally disables the quantile lines by setting `quantiles = NULL`.
To make this clearer, I have added more descriptive comments to the test file to clarify the purpose of each test case.

d-morrison replied:
that wasn't what this test did before; look at how the snapshot has changed after this edit:

https://github.com/UCD-SERG/serocalculator/pull/455/files#diff-f11a0031a2784fb17656d3c3831e8d78d76eea2245bf14e0a5f75a0d2a16d4ae

I see that `show_quantiles = TRUE` was out of date; but instead of replacing it with `quantiles = NULL`, how about just deleting that line? Then I think the snapshot will go back to how it was before?

```suggestion
```

Kwan-Jenny replied:
Oh, that was my mistake. I thought it was from the first chunk of code, the plot3 code lines. Thanks for pointing that out!

d-morrison commented on R/graph.curve.params.R:
```diff
@@ -151,16 +155,30 @@ graph.curve.params <- function( # nolint: object_name_linter
+            data = sc_to_graph,
+            alpha = alpha_samples,
+            aes(
+              color = .data$chain |> factor(),
```
I think we can avoid the if-else with this approach:

```suggestion
              color = if (chain_color) .data$chain |> factor(),
```

here, color = "black" is default so the `else` doesn't need to specified.

Kwan-Jenny replied:
@d-morrison I changed the whole chunk of this code. Am I correctly following your suggestion?

d-morrison replied:
@Kwan-Jenny not quite; there's a simpler approach. revert that commit and start a teams call and I'll show you

d-morrison commented on R/graph.curve.params.R:
```diff
@@ -202,8 +220,15 @@ graph.curve.params <- function( # nolint: object_name_linter
+          ggplot2::labs(col = "")
+      }
+    } else {
+      plot1 <- plot1 + ggplot2::labs(col = "Quantile")
```
redundant? 
<img width="241" height="24" alt="image" src="https://github.com/user-attachments/assets/061b4059-ea4c-43f1-ad0e-160275532bee" />

Kwan-Jenny replied:
@d-morrison Fixed!

d-morrison reviewed (changes requested):
Looks good! a few notes

Kwan-Jenny committed:
Update R/graph.curve.params.R

Co-authored-by: Douglas Ezra Morrison <demorrison@ucdavis.edu>

Kwan-Jenny committed:
# feat: Refactor and improve graph.curve.params

This commit refactors the graph.curve.params function for clarity and improved output.

The if/else block for adding geom_line layers has been streamlined by dynamically building a parameter list. This makes the code more concise and easier to maintain.

Additionally, the legend title 'Quantile' has been removed when only quantile lines are plotted, resulting in a cleaner visualization.

Kwan-Jenny committed:
# test(graph.curve.params): Improve test case descriptions

This commit adds more descriptive comments to the tests for graph.curve.params to clarify the purpose of each test case. This makes it clear why certain parameters are used, for example, to specifically test the disabling of quantile lines.

Kwan-Jenny committed:
Revert "# feat: Refactor and improve graph.curve.params"

This reverts commit bccf3119cf5a2056385e75f6e389446956d47ffb.

Kwan-Jenny committed:
# refactor(graph.curve.params): Simplify geom_line call

This commit refactors the plotting logic for MCMC samples by consolidating the if/else block into a single geom_line call. A conditional is now used inside the aes() mapping to dynamically handle the color aesthetic, making the code more concise and removing duplication.

Kwan-Jenny committed:
removing blank lines

Kwan-Jenny committed:
Revert "removing blank lines"

This reverts commit 760a6c5142937581346d8f7ac37031fd87acd46e.

d-morrison commented on R/graph.curve.params.R:
```diff
@@ -202,8 +205,15 @@ graph.curve.params <- function( # nolint: object_name_linter
+          ggplot2::labs(col = "")
+      }
+    } else {
+      plot1 <- plot1 + ggplot2::labs(col = "")
```
combine these if-elses?

```
    if ((length(iters_to_graph) > 0) && chain_color) {
      plot1 <- plot1 +
        ggplot2::labs(col = "MCMC chain")
    } else {
      plot1 <- plot1 +
        ggplot2::labs(col = "")
    }
```

Kwan-Jenny replied:
@d-morrison Thanks for the suggestion to combine the if-else statements. I used that idea to refactor the code by first determining the label text with a single conditional, and then adding it to the plot in one step. This avoids repeating the ggplot2::labs() call and makes the code's intent a bit clearer. I slightly edited your suggested code.

d-morrison replied:
@Kwan-Jenny That's a great improvement - I love it!!

d-morrison reviewed (changes requested):
almost there! two more comments

Kwan-Jenny committed:
# refactor(graph.curve.params): Simplify legend label logic

This commit refactors the nested if-else block for setting the legend label. The logic is now consolidated into a single conditional expression to determine the label text, which is then applied in one step. This improves readability and maintainability by removing duplicated code.

Kwan-Jenny committed:
# minor revision

d-morrison reviewed (approved):
Thanks! This PR looks great to me - short and sweet!
